ngetopt.2
.\" Man page generated from reStructuredText.
.
.TH NGETOPT.AWK 2 "" "0.1" "gawk library"
.SH NAME
ngetopt.awk \- AWK-Library for command line parsing
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
A command line option parsing library for \fIgawk4\fP written in \fIgawk4\fP\&.  Abolish the need to write shell command line parsers for the small programs.  Typical use looks like this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
#!/usr/bin/gawk \-E
@include "ngetopt"
BEGIN {
    regopt(opstring)
    parseopt()
    usage()
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
A line\-by\-line breakdown:
.INDENT 0.0
.TP
.B #!/usr/bin/gawk \-E
The \fI\-E\fP\-switch tells \fIgawk\fP to disable its own command line processing, so we have the line all to ourselves.
.TP
.B @include "ngetopt"
Attempts to find \fIngetopt\fP or \fIngetopt.awk\fP the awk library path. See \fIFILES\fP for more information.
.UNINDENT
.sp
After setup, the following user accessible functions are available:
.INDENT 0.0
.TP
.B regopt()
A convenience function to add a new entry to the globally available \fIopt\fP\-array.
.TP
.B parseopt()
Parses the command line, making use of array \fIopt\fP\&. Will set configuration as needed.
.TP
.B usage()
With the function \fIusage()\fP, a help message is generated from the defined options.
.UNINDENT
.SH DESCRIPTION
.sp
Central to \fIngetopt.awk\fP is the array \fIopt\fP\&.
.sp
The \fIopt\fP\-array needs to have all information of the options.  There is a numeric index for each option, for which a few properties shall be defined.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
opt[1]["short"]="F"
opt[1]["long"]="field\-separator"
opt[1]["flag"]="FS"
opt[1]["description"]="Sets the field separator."
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Properties of the \fIopt\fP\-array that have special meaning are the following:
.INDENT 0.0
.TP
.B short
The short option. Will fail if it is more than one character.
.TP
.B long
The long option. Will be parsed with \fB\-\-\fP before it.
.UNINDENT
.sp
A registered option requires one short option, one long option, or a long option and a short option.
.INDENT 0.0
.TP
.B has_arg
\fIRequired\fP Defines the type of option. Required. Accepted values:
\fB0\fP or \fBno\fP: Option does not require an argument
\fB1\fP or \fByes\fP: Option requires an argument
\fB2\fP or \fBmaybe\fP: Not implemented.
Required.
.TP
.B flag
The variable that will contain whatever has been specified on the command line. This is what your program will need. Required.
.TP
.B desc
Description. Used when displaying the \fBusage()\fP\-function. Optional.
.TP
.B default
The value \fBflag\fP will assume if specified.  As \fIoptional arguments\fP are as of yet not supported, use is limited to shortcuts.
.TP
.B vals
This defines the permissable values of the option.  A \fB|\fP\-separated string serves as a definition of possibly values.
.UNINDENT
.SS regopt()
.sp
In stead of constructing the \fIopt\fP\-array manually, the utility function \fIregopt\fP can be used.  For each call, it adds a new option to the \fIopt\fP\-array.  An example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
regopt("short=F;long=field\-separator;has_arg=yes;desc=Define the Field Separator")
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fIregopt\fP accepts one \fI;\fP\-separated string of \fI=\fP\-separated key\-value pairs.  Permissable keys are listed in the previous section.
.SH FILES
.sp
\fIgawk\fP version 4 should be available.  The library has been tested with the current stable release, \fIgawk 4.1.1\fP\&.
Put \fInregopt.awk\fP in \fI$AWKPATH\fP\&.  If you did not set \fI$AWKPATH\fP as an environment variable in your shell startup scripts, you can establish the location like this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
%\ gawk \(aqBEGIN{print ENVIRON["AWKPATH"]}\(aq
\&.:/usr/share/awk
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
On my system, \fIgawk\fP will first look in the current directory for a file to include, then in \fI/usr/share/awk\fP\&.
.SH EXAMPLES
.sp
See the \fBtest/\fP\-directory for some examples.  Also see \fI\%table\fP for a real world application.
.SH BUGS
.sp
Probably.  Send a pull request or open an issue on \fI\%https://github.com/joepvd/ngetopt.awk\fP
.SH SEE ALSO
.sp
\fIgetopt(3)\fP, \fIgawk(1)\fP
.SH AUTHOR
joepvd
.SH COPYRIGHT
GNU
.\" Generated by docutils manpage writer.
.


ngetopt.awk
# Library for option parsing with gawk programs. Original program
# needs to be started with the `-E` option, which disables all the
# option processing of gawk itself. 
#
# By joepvd, 2015
# https://github.com/joepvd/ngetopt.awk
# 
# Usage: 
# #!/usr/bin/gawk -E
# @include "noptparse.awk"
# BEGIN {
#     ...
# }
#
# DONE: -   Reimplement "Default".
#           Use case: To allow for short cuts that use the same flag. From
#           `table`: --rst should be equivalent to --style rst.
# DONE: -   Implement "possible values"
#           Keep the logic of checking the permissable values in one 
#           centralized place. 
# TODO: -   Consider implementing "leave as file"-flag
#           Would that make sense? Or would it be recommendable to have
#           the user fix it for concrete cases?  As the order is not 
#           guaranteed... Maybe recommend `FILENAME == file` and 
#           `FILENAME != file`-blocks. 
# TODO: -   Optional arguments

function regopt(str,        n, pairs, keyval, i, j, o, key, val, s, v, k, pv) {
    # Convenience function to register option parameters into the globally
    # available `opt`-array. Takes a string as argument that is colon separated 
    # list of key-value pairs. Keys and values are separated with a `=`-sign. 

    # TODO: Commentify the variable reservation in the function declaration. 

    if (1 in opt) {} # Now, opt is an array, even if it did not exist before. 
    n=length(opt)+1  # Index of the soon-to-be registered option.

    split(str, pairs, "[[:space:]]*;[[:space:]]*")
    for (o in pairs) {
        split(pairs[o], keyval, "=")
        key = keyval[1]
        val = keyval[2]
        if (key == "has_arg") {
            if (val ~ /^(yes|Y|y|YES|0)$/)
                opt[n]["has_arg"] = "0"
            else
                opt[n]["has_arg"] = "1" 
        } 
        else if (key == "vals") {
            split(val, pv, "|")
            for (k in pv) {
                #print "k", k
                opt[n]["vals"][pv[k]]
            }
        }
        else opt[n][key] = val
    }
    # Registered the input.
    # Check if the obligatory stuff is available: 
    if ( ! "long" in opt[n] && ! "short" in opt[n] ) {
        printf "noptparse: Neither long nor short option " >"/dev/stderr"
        printf "detected. Terminating.\n" >"/dev/stderr"
        _assert_exit = 1
        exit 1
    } else if ( ! "dest" in opt[n] ) {
        printf("No destination for the command line option.\n")>"/dev/stderr"
        _assert_exit = 1
        exit 1
    } else if ( ! "has_arg" in opt[n] ) {
        printf("Specify `has_arg` for option.\n")>"/dev/stderr"
        _assert_exit = 1
        exit 1
    }
    # Apparently, all is fine now. 
    # For those interested, what exactly has been achieved? 
    if ( opt_debug == "y") {
        for ( key in opt[n] ) {
            s = sprintf("ngetopt: regopt: opt[\"%s\"][\"%s\"] = ", n, key)
            if ( isarray(opt[n][key]) ) {
                for (v=1;v<=length(opt[n][key]);v++) {
                    printf "%s\"%s\"\n", s, opt[n][key][v] >"/dev/stderr"
                }
            } else {
                printf "%s\"%s\"\n", s, opt[n][key] >"/dev/stderr"
            }
        }
    }
}

function parseopt(opt,          i, nextopt, opt_flag, argind, shopt, argopt, 
                                optstring, f, c, out) {
    # TODO: Commentify the variable reservation in the function declaration. 
    if (opt_debug == "y") {
        printf("ngetopt: parseopt: Processing command line: ARGV[%s]=\"%s\"",
               0, ARGV[0]) >"/dev/stderr"
        for (i=1;i<=ARGC-1;i++) {
            printf "; ARGV[%s]=\"%s\"", i, ARGV[i] >"/dev/stderr"
        }
        printf ".\n" >"/dev/stderr"
    }
    argind=1
    if (1 in config) {} # Is an array now...
                        # This is a globally available array for debugging/
                        # testing purposes.
    while (argind in ARGV) {
        #printf "argind: %s ARGV[argind] = %s\n", argind, ARGV[argind]>"/dev/stderr"
        argopt="n"      # registers whether the argument comes from `argind+1`.
        if (opt_debug == "y") {
            printf("ngetopt: parseopt: ARGV[%s]=\"%s\": ",
                   argind, ARGV[argind]) >"/dev/stderr"
        }
        if (ARGV[argind] ~ /^(--|[^-].*)$/) {
            if (opt_debug == "y") {
                printf("No option. The rest left as file.\n",
                       ARGV[argind]) >"/dev/stderr"
            }
            if (ARGV[argind] == "--") {
                delete ARGV[argind]
            }
            break
        } else if (ARGV[argind] ~ /^-[[:alpha:]]/) {
            # A short option! Some juggling with concatenated short options. 
            optstring=substr(ARGV[argind], 2, length(ARGV[argind]))
            while (length(optstring)>0) {
                # Looping over the chars in optstring, using "shopt"
                # (short option) as index.
                shopt = substr(optstring, 1, 1)
                if (length(optstring)==1) {
                    nextopt = "--"
                    # Overriding previous assignment if necessary:
                    if (argind+1 in ARGV) {
                        nextopt = ARGV[argind+1]
                        argopt="y"
                    }
                } else {
                    # If argument is needed, it will come from the rest of the
                    # current shell word.
                    nextopt = substr(optstring, 2, length(optstring))
                }
                opt_flag = getopt(shopt, nextopt)
                if (opt_flag == "file") {
                    printf("ngetopt: parseopt: Unknown option %s, exiting.\n",
                           shopt) >"/dev/stderr"
                    _assert_exit = 1
                    exit 1
                } else if (opt_flag == "opt_noarg") {
                    if (nextopt == "--" || length(optstring)==1) {
                        if (argind in ARGV){
                            delete ARGV[argind]
                            argind++
                        }
                        break
                    } else {
                        optstring = nextopt
                    }
                } else if (opt_flag == "opt_arg") {
                    delete ARGV[argind]
                    argind+=1
                    optstring=""
                    if (argopt == "y") {
                        # The argument came from the next `ARGV`. 
                        delete ARGV[argind]
                        argind+=1
                    }
                    break
                } else {
                    # We really should not be here. 
                    printf("Illegal state. Exiting\n") >"/dev/stderr"
                    _assert_exit = 1
                    exit 1
                }
            }
        } else {
            # A long option. 
            if (argind+1 in ARGV) {
                nextopt = ARGV[argind+1]
            } else {
                nextopt = "--"
            }
            opt_flag = getopt(ARGV[argind], nextopt)
            if (opt_flag == "file") {
                if (opt_debug == "y") {
                    printf("No option. The rest left as file.\n",
                           ARGV[argind]) >"/dev/stderr"
                }
                break
            }
            else if (opt_flag == "opt_noarg") {
                delete ARGV[argind]
                argind+=1
            }
            else if (opt_flag == "opt_arg") {
                delete ARGV[argind]
                delete ARGV[argind+1]
                argind+=2
            }
        continue
        }
    }

    # Done processing. Tricking the gawk syntax parser into picking
    # up possible changes to user modifiable variables that awk 
    # explicitly needs to be aware of: 

    BINMODE = SYMTAB["BINMODE"]
    CONVFMT = SYMTAB["CONVFMT"]
    FIELDWIDTHS = SYMTAB["FIELDWIDTHS"]
    FPAT = SYMTAB["FPAT"]
    FS = SYMTAB["FS"]
    IGNORECASE = SYMTAB["IGNORECASE"]
    LINT = SYMTAB["LINT"]
    OFMT = SYMTAB["OFMT"]
    OFS = SYMTAB["OFS"]
    ORS = SYMTAB["ORS"]
    PREC = SYMTAB["PREC"]
    ROUNDMODE = SYMTAB["ROUNDMODE"]
    RS = SYMTAB["RS"]
    SUBSEP = SYMTAB["SUBSEP"]
    TEXTDOMAIN = SYMTAB["TEXTDOMAIN"]
    
    # Printing a summary if desired: 
    if (opt_debug ~ /^(y|t)$/) {
        if (opt_debug == "t") { out = "/dev/stdout" }
        else { out = "/dev/stderr" }
        for (c in config) {
            printf("ngetopt: parseopt: result: option: %s = \"%s\"\n",
                    c, SYMTAB[c]) >out
        }
        for (f in ARGV) {
            if (f>0) {
                printf("ngetopt: parseopt: result: file: \x27%s\x27\n",
                       ARGV[f]) >out
            }
        }
        if (length(ARGV)=="1") {
            printf("ngetopt: parseopt: result: file: <NONE>\n")>out
        }
    }
}

function getopt(option, nextopt,            o, flag, opt_flag, val, failstr, i, pval) {
    opt_flag="file"
    for ( o in opt ) {
        if (option != "-"opt[o]["short"] && \
            option != "--"opt[o]["long"] && \
            option != opt[o]["short"]) {
                continue
        }
        if (opt[o]["has_arg"] == "1") {
            # Option has no argument
            flag = opt[o]["flag"]
            val = opt[o]["val"]==""?"yes":opt[o]["val"]
            SYMTAB[flag] = val
            if (opt_debug ~ /^(y|t)$/) { config[flag]=1 }
            if (opt_debug == "y") {
                printf("%s = \"%s\"\n",
                       flag, SYMTAB[flag]) >"/dev/stderr"
            }
            opt_flag="opt_noarg"
            break
        } else if (opt[o]["has_arg"] == "0") {
            # Option needs argument, verify it has one. 
            flag = opt[o]["flag"]
            if (nextopt ~ /^(|--)$/) {
                printf("Expecting argument to `%s`, none received. Exiting.\n", 
                       option) >"/dev/stderr"
                _assert_exit = 1
                exit 1
            }
            if ("vals" in opt[o]){
                # Only some values are allowed. 
                # Check if the provided value is one of the possible ones. 
                pval = 1
                for (i in opt[o]["vals"]) {
                    failstr=sprintf("%s%s", failstr==""?"":failstr", ", "«"i"»")
                    if (nextopt == i) {
                        pval = 0
                        continue
                    }
                }
                if (pval == 1) {
                    printf("Argument «%s» expects one of the following values: %s. ",
                           option, failstr)>"/dev/stderr"
                    printf("Received: «%s». Exiting.\n", nextopt) >"/dev/stderr"
                    _assert_exit = 1
                    exit 1
                }
            }
            SYMTAB[flag] = nextopt
            if (opt_debug ~ /^(y|t)$/) { config[flag]=1 }
            opt_flag="opt_arg"
            if (opt_debug == "y") {
                printf("%s = \"%s\"\n", flag, SYMTAB[flag]) >"/dev/stderr"
            }
            break
        } 
    }
    return opt_flag
}

function usage(         p,n) {
    # TODO: Some nice formatting with columns...
    # TODO: Generate Usage according to GNU-specs. 
    split(ENVIRON["_"], p, "/")
    printf "\nUsage of %s:\n\n", p[length(p)] >"/dev/stderr"
    for (n in opt) {
        printf("    -%s (--%s): %s\n",
               opt[n]["short"], opt[n]["long"], opt[n]["desc"]) >"/dev/stderr"
    }
}

END {
    # Need an END{}-block, as the exit in an included file does not get
    # propagated to the program, and the program will be waiting for input.
    if (_assert_exit)
        exit 1
}

